client/ClientDecisionService/InMemoryLogger.cs

```diff
@@ -0,0 +1,215 @@
+        /// of a given event. In other words, the <x,a,r,p> tuple.
+        /// </summary>
+        internal class DataPoint : IEvent
+        {
+            public string Key { get; set; }
+            // Contains the context, action, and probability (at least)
+            public Interaction InteractData { get; set; }
+            public float Reward { get; set; }
+            // TODO: This can be used to support custom reward functions
+            public ConcurrentBag<object> Outcomes = new ConcurrentBag<object>();
+            // Used to control expiration for fixed-duration events
+            public DateTime ExpiresAt = DateTime.MaxValue;
+        }
+
+        // The experimental unit duration, or how long to wait for reward information before 
+        // completing an event (TimeSpan.MaxValue means wait forever)
+        private TimeSpan experimentalUnit;
+        // Stores pending (incomplete) events, either for a fixed experimental unit duration or for 
+        // manual completion
+        private ConcurrentDictionary<string, DataPoint> pendingData;
+        // A queue and timer to expire events based on the experimental unit duration
+        private ConcurrentQueue<DataPoint> completionQueue;
+        private Timer completionTimer;
+        // Stores expired or manually-completed events 
+        private ConcurrentDictionary<string, DataPoint> completeData = new ConcurrentDictionary<string, DataPoint>();
+        // If no reward information is received, this value will be used
+        private float defaultReward;
+
+        /// <summary>
+        /// Creates a new in-memory logger for exploration data
+        /// </summary>
+        /// <param name="expUnit">The experimental unit duration, or how long to wait for reward 
+        /// information. Set this to TimeSpan.MaxValue for infinite duration (events never expire
+        /// and must be completed manually.</param>
+        /// <param name="defaultReward">Reward value to use when no reward signal is received</param>
+        public InMemoryLogger(TimeSpan expUnit, float defaultReward = (float)0.0)
+        {
+            this.experimentalUnit = expUnit;
+            this.defaultReward = defaultReward;
+            pendingData = new ConcurrentDictionary<string,DataPoint>();
+            // We only need the completion queue/timer if events are being completed automatically
+            // (by experimental duration)
+            if (experimentalUnit != TimeSpan.MaxValue)
+            {
+                completionQueue = new ConcurrentQueue<DataPoint>();
+                completionTimer = new Timer(experimentalUnit.TotalMilliseconds);
+                completionTimer.Elapsed += completeExpiredEvents;
+                completionTimer.AutoReset = false;
+                completionTimer.Start();
+            }
+        }
+
+        private void completeExpiredEvents(object sender, ElapsedEventArgs e)

```

Comment:
Yes, I believe its possible to use Rx. However to keep things simple I think we should avoid the dependency for now; the Berkeley folks and possibly others will likely need to use DecisionServiceLocal in a different language (e.g. C++ or Rust), so having all the relevant functionality clearly visible will enable such translation if it is needed.